File: 547.friend-circles.java
public class Solution {
    public int findCircleNum(int[][] M) {
        //DFS Solution 
        boolean[] visited = new boolean[M.length];
        int count = 0;
        for(int i = 0; i < M.length; i++) 
            if(!visited[i]) {
                dfs(M, visited, i);
                count++;
            }
        return count;
    }
    private void dfs(int[][] M, boolean[] visited, int i) {
        for(int j = 0; j < M.length; j++) 
            if(M[i][j] == 1 && !visited[j]) {
                visited[j] = true;
                dfs(M, visited, j);
            }
    }
}

File: 547.friend-circles.js
/*
 * @lc app=leetcode id=547 lang=javascript
 *
 * [547] Friend Circles
 */
/**
 * @param {number[][]} M
 * @return {number}
 */
var findCircleNum = function (M) {

  if (M.length === 0) {
    return 0;
  }

  let len = M.length;
  let result = 0;

  for (let i = 0; i < len; i++) {
    for (let j = 0; j < M[0].length; j++) {
      if (M[i][j] === 1) {
        result++;
        findFriends(M, i, j);
      }
    }
  }


  return result;
};

var findFriends = function (M, i, j) {

  if (M[i][j] !== 1) {
    return;
  }
  //mark it visited
  M[i][j] = '#';

  for (let k = 0; k < M.length; k++) {
    findFriends(M, j, k);
  }
};

